### DISCUSSION TITLE
package `solana-program v1.18.8` cannot be built

### DISCUSSION THREAD

**Discussion by Zero9BSC** [otter-sec/anchor#2866] [Q&A]
zero9@DESKTOP-BHNNLI6:~/poi$ anchor build
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
error: package `solana-program v1.18.8` cannot be built because it requires rustc 1.75.0 or newer, while the currently active rustc version is 1.72.0-dev
Either upgrade to rustc 1.75.0 or newer, or use
cargo update -p solana-program@1.18.8 --precise ver
where `ver` is the latest version of `solana-program` supporting rustc 1.72.0-dev
zero9@DESKTOP-BHNNLI6:~/poi$ curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
info: downloading installer

Welcome to Rust!

This will download and install the official compiler for the Rust
programming language, and its package manager, Cargo.

Rustup metadata and toolchains will be installed into the Rustup
home directory, located at:

  /home/zero9/.rustup

This can be modified with the RUSTUP_HOME environment variable.

The Cargo home directory is located at:

  /home/zero9/.cargo

This can be modified with the CARGO_HOME environment variable.

The cargo, rustc, rustup and other commands will be added to
Cargo's bin directory, located at:

  /home/zero9/.cargo/bin

This path will then be added to your PATH environment variable by
modifying the profile files located at:

  /home/zero9/.profile
  /home/zero9/.bashrc

You can uninstall at any time with rustup self uninstall and
these changes will be reverted.

Current installation options:

   default host triple: x86_64-unknown-linux-gnu
     default toolchain: stable (default)
               profile: default
  modify PATH variable: yes

1) Proceed with standard installation (default - just press enter)
2) Customize installation
3) Cancel installation
>1

info: profile set to 'default'
info: default host triple is x86_64-unknown-linux-gnu
warning: Updating existing toolchain, profile choice will be ignored
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: default toolchain set to 'stable-x86_64-unknown-linux-gnu'

  stable-x86_64-unknown-linux-gnu unchanged - rustc 1.77.0 (aedd173a2 2024-03-17)

Rust is installed now. Great!

To get started you may need to restart your current shell.
This would reload your PATH environment variable to include
Cargo's bin directory ($HOME/.cargo/bin).

To configure your current shell, you need to source
the corresponding env file under $HOME/.cargo.

This is usually done by running one of the following (note the leading DOT):
. "$HOME/.cargo/env"            # For sh/bash/zsh/ash/dash/pdksh
source "$HOME/.cargo/env.fish"  # For fish
zero9@DESKTOP-BHNNLI6:~/poi$ anchor build
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
error: package `solana-program v1.18.8` cannot be built because it requires rustc 1.75.0 or newer, while the currently active rustc version is 1.72.0-dev
Either upgrade to rustc 1.75.0 or newer, or use
cargo update -p solana-program@1.18.8 --precise ver
where `ver` is the latest version of `solana-program` supporting rustc 1.72.0-dev
zero9@DESKTOP-BHNNLI6:~/poi$ source "$HOME/.cargo/env"
zero9@DESKTOP-BHNNLI6:~/poi$ rustc --version
rustc 1.77.0 (aedd173a2 2024-03-17)
zero9@DESKTOP-BHNNLI6:~/poi$ anchor build
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
error: package `solana-program v1.18.8` cannot be built because it requires rustc 1.75.0 or newer, while the currently active rustc version is 1.72.0-dev
Either upgrade to rustc 1.75.0 or newer, or use
cargo update -p solana-program@1.18.8 --precise ver
where `ver` is the latest version of `solana-program` supporting rustc 1.72.0-dev
zero9@DESKTOP-BHNNLI6:~/poi$ rm -rf target
zero9@DESKTOP-BHNNLI6:~/poi$ anchor build
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
error: package `solana-program v1.18.8` cannot be built because it requires rustc 1.75.0 or newer, while the currently active rustc version is 1.72.0-dev
Either upgrade to rustc 1.75.0 or newer, or use
cargo update -p solana-program@1.18.8 --precise ver
where `ver` is the latest version of `solana-program` supporting rustc 1.72.0-dev
zero9@DESKTOP-BHNNLI6:~/poi$ anchor build
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
error: package `solana-program v1.18.8` cannot be built because it requires rustc 1.75.0 or newer, while the currently active rustc version is 1.72.0-dev
Either upgrade to rustc 1.75.0 or newer, or use
cargo update -p solana-program@1.18.8 --precise ver
where `ver` is the latest version of `solana-program` supporting rustc 1.72.0-dev
zero9@DESKTOP-BHNNLI6:~/poi$ anchor build
warning: both `/home/zero9/poi/rust-toolchain` and `/home/zero9/poi/rust-toolchain.toml` exist. Using `/home/zero9/poi/rust-toolchain`
info: syncing channel updates for '1.77.0-x86_64-unknown-linux-gnu'
info: latest update on 2024-03-21, rust version 1.77.0 (aedd173a2 2024-03-17)
info: downloading component 'cargo'
info: downloading component 'clippy'
info: downloading component 'rust-docs'
info: downloading component 'rust-std'
info: downloading component 'rustc'
info: downloading component 'rustfmt'
info: installing component 'cargo'
info: installing component 'clippy'
info: installing component 'rust-docs'
info: installing component 'rust-std'
info: installing component 'rustc'
info: installing component 'rustfmt'
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
error: package `solana-program v1.18.8` cannot be built because it requires rustc 1.75.0 or newer, while the currently active rustc version is 1.72.0-dev
Either upgrade to rustc 1.75.0 or newer, or use
cargo update -p solana-program@1.18.8 --precise ver
where `ver` is the latest version of `solana-program` supporting rustc 1.72.0-dev
zero9@DESKTOP-BHNNLI6:~/poi$ rustup override set 1.77.0
info: override toolchain for '/home/zero9/poi' set to '1.77.0-x86_64-unknown-linux-gnu'
zero9@DESKTOP-BHNNLI6:~/poi$ anchor build
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
error: package `solana-program v1.18.8` cannot be built because it requires rustc 1.75.0 or newer, while the currently active rustc version is 1.72.0-dev
Either upgrade to rustc 1.75.0 or newer, or use
cargo update -p solana-program@1.18.8 --precise ver
where `ver` is the latest version of `solana-program` supporting rustc 1.72.0-dev
zero9@DESKTOP-BHNNLI6:~/poi$ cargo build
warning: virtual workspace defaulting to `resolver = "1"` despite one or more workspace members being on edition 2021 which implies `resolver = "2"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
note: for more details see https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions
   Compiling proc-macro2 v1.0.79
   Compiling unicode-ident v1.0.12
   Compiling version_check v0.9.4
   Compiling serde v1.0.197
   Compiling typenum v1.17.0
   Compiling cfg-if v1.0.0
   Compiling libc v0.2.153
   Compiling subtle v2.5.0
   Compiling syn v1.0.109
   Compiling autocfg v1.1.0
   Compiling once_cell v1.19.0
   Compiling thiserror v1.0.58
   Compiling cpufeatures v0.2.12
   Compiling ppv-lite86 v0.2.17
   Compiling anyhow v1.0.81
   Compiling semver v1.0.22
   Compiling serde_json v1.0.114
   Compiling generic-array v0.14.7
   Compiling num-traits v0.2.18
   Compiling ryu v1.0.17
   Compiling itoa v1.0.10
   Compiling num-bigint v0.4.4
   Compiling unicode-segmentation v1.11.0
   Compiling either v1.10.0
   Compiling ahash v0.8.11
   Compiling crossbeam-utils v0.8.19
   Compiling getrandom v0.1.16
   Compiling bs58 v0.5.1
   Compiling heck v0.3.3
   Compiling rustc_version v0.4.0
   Compiling quote v1.0.35
   Compiling crunchy v0.2.2
   Compiling paste v1.0.14
   Compiling zerocopy v0.7.32
   Compiling proc-macro-error-attr v1.0.4
   Compiling wasm-bindgen-shared v0.2.92
   Compiling syn v2.0.53
   Compiling itertools v0.10.5
   Compiling proc-macro-error v1.0.4
   Compiling equivalent v1.0.1
   Compiling hashbrown v0.14.3
   Compiling log v0.4.21
   Compiling rayon-core v1.12.1
   Compiling crossbeam-epoch v0.9.18
   Compiling hashbrown v0.13.2
   Compiling crossbeam-deque v0.8.5
   Compiling ahash v0.7.8
   Compiling bumpalo v3.15.4
   Compiling feature-probe v0.1.1
   Compiling getrandom v0.2.12
   Compiling jobserver v0.1.28
   Compiling rand_core v0.6.4
   Compiling num-integer v0.1.46
   Compiling indexmap v2.2.6
   Compiling winnow v0.5.40
   Compiling toml_datetime v0.6.5
   Compiling cc v1.0.90
   Compiling rand_core v0.5.1
   Compiling rand_chacha v0.3.1
   Compiling bv v0.11.1
   Compiling rand_chacha v0.2.2
   Compiling rand v0.8.5
   Compiling bitmaps v2.1.0
   Compiling solana-frozen-abi-macro v1.18.8
   Compiling im v15.1.0
   Compiling cfg_aliases v0.1.1
   Compiling opaque-debug v0.3.1
   Compiling rustversion v1.0.14
   Compiling tinyvec_macros v0.1.1
   Compiling tinyvec v1.6.0
   Compiling borsh v1.3.1
   Compiling rayon v1.9.0
   Compiling ark-std v0.4.0
   Compiling sized-chunks v0.6.5
   Compiling rand v0.7.3
   Compiling rand_xoshiro v0.6.0
   Compiling solana-frozen-abi v1.18.8
   Compiling memoffset v0.9.0
   Compiling bs58 v0.4.0
   Compiling blake3 v1.5.1
   Compiling wasm-bindgen v0.2.92
   Compiling arrayref v0.3.7
   Compiling toml_edit v0.21.1
   Compiling unicode-normalization v0.1.23
   Compiling solana-program v1.18.8
   Compiling hashbrown v0.11.2
   Compiling memmap2 v0.5.10
   Compiling arrayvec v0.7.4
   Compiling constant_time_eq v0.3.0
   Compiling lazy_static v1.4.0
   Compiling byteorder v1.5.0
   Compiling rustc-hash v1.1.0
   Compiling base64 v0.12.3
   Compiling keccak v0.1.5
   Compiling base64 v0.21.7
   Compiling base64 v0.13.1
   Compiling borsh-derive-internal v0.10.3
   Compiling borsh-schema-derive-internal v0.10.3
   Compiling borsh-derive-internal v0.9.3
   Compiling borsh-schema-derive-internal v0.9.3
   Compiling wasm-bindgen-backend v0.2.92
   Compiling proc-macro-crate v3.1.0
   Compiling ark-serialize-derive v0.4.2
   Compiling derivative v2.2.0
   Compiling ark-ff-macros v0.4.2
   Compiling ark-ff-asm v0.4.2
   Compiling anchor-derive-space v0.29.0
   Compiling serde_derive v1.0.197
   Compiling thiserror-impl v1.0.58
   Compiling zeroize_derive v1.4.2
   Compiling syn_derive v0.1.8
   Compiling bytemuck_derive v1.6.0
   Compiling wasm-bindgen-macro-support v0.2.92
   Compiling num-derive v0.4.2
   Compiling solana-sdk-macro v1.18.8
   Compiling borsh-derive v1.3.1
   Compiling zeroize v1.3.0
   Compiling wasm-bindgen-macro v0.2.92
   Compiling bytemuck v1.15.0
   Compiling toml v0.5.11
   Compiling serde_bytes v0.11.14
   Compiling bitflags v2.5.0
   Compiling bincode v1.3.3
   Compiling crypto-common v0.1.6
   Compiling block-buffer v0.10.4
   Compiling digest v0.9.0
   Compiling crypto-mac v0.8.0
   Compiling block-buffer v0.9.0
   Compiling pbkdf2 v0.4.0
   Compiling libsecp256k1-core v0.2.2
   Compiling hmac v0.8.1
   Compiling sha2 v0.9.9
   Compiling curve25519-dalek v3.2.1
   Compiling digest v0.10.7
   Compiling hmac-drbg v0.3.0
   Compiling sha2 v0.10.8
   Compiling ark-serialize v0.4.2
   Compiling sha3 v0.10.8
   Compiling tiny-bip39 v0.8.2
   Compiling ark-ff v0.4.2
   Compiling anchor-syn v0.29.0
   Compiling proc-macro-crate v0.1.5
   Compiling libsecp256k1-gen-ecmult v0.2.1
   Compiling libsecp256k1-gen-genmult v0.2.1
   Compiling borsh-derive v0.10.3
   Compiling borsh-derive v0.9.3
   Compiling libsecp256k1 v0.6.0
   Compiling borsh v0.10.3
   Compiling borsh v0.9.3
   Compiling anchor-attribute-error v0.29.0
   Compiling anchor-attribute-event v0.29.0
   Compiling anchor-attribute-program v0.29.0
   Compiling anchor-derive-serde v0.29.0
   Compiling anchor-attribute-account v0.29.0
   Compiling anchor-attribute-access-control v0.29.0
   Compiling anchor-derive-accounts v0.29.0
   Compiling anchor-attribute-constant v0.29.0
   Compiling ark-poly v0.4.2
   Compiling ark-ec v0.4.2
   Compiling ark-bn254 v0.4.0
   Compiling light-poseidon v0.2.0
   Compiling anchor-lang v0.29.0
   Compiling poi v0.1.0 (/home/zero9/poi/programs/poi)
warning: unused variable: `ctx`
 --> programs/poi/src/lib.rs:9:23
  |
9 |     pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
  |                       ^^^ help: if this is intentional, prefix it with an underscore: `_ctx`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: `poi` (lib) generated 1 warning (run `cargo fix --lib -p poi` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 27.13s

**Comment by Anujoo9** [comment]
Try this it will work

Edit cargo toml
```[dependencies]

    borsh = ">=0.9.3"  # Update to the latest version compatible with your other dependencies
    borsh-derive = ">=0.10.0"  # Update to the latest version compatible with your other dependencies
    solana-program = ">=1.18.17"  # Specify a version range that includes the latest version
    solana-program-test = ">=1.18.17"  # Specify a version range that includes the latest version
    solana-sdk = ">=1.18.17"  # Specify a version range that includes the latest version
```
#then run
cargo update
